// Protocol messages for program graphs.
//
// Copyright 2019-2020 the ProGraML authors.
//
// Contact Chris Cummins <chrisc.101@gmail.com>.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package programl;

option go_package = "programlpb";
option java_multiple_files = true;
option java_outer_classname = "ProgramGraphProto";
option java_package = "com.programl";

import "programl/proto/edge.proto";
import "programl/proto/features.proto";
import "programl/proto/function.proto";
import "programl/proto/module.proto";
import "programl/proto/node.proto";

// A program graph.
//
// Module -> Function -> Node
//
// Edges connect nodes, without an explicit scope.
message ProgramGraph {
  repeated Node node = 1;
  repeated Edge edge = 2;
  repeated Function function = 4;
  repeated Module module = 5;
  Features features = 6;
}

// A list of program graphs.
message ProgramGraphList {
  Features context = 1;
  repeated ProgramGraph graph = 2;
}
